routing/turns_tts_text_i18n.cpp

```diff
@@ -0,0 +1,187 @@
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+#include <regex>
+#include <string>
+
+namespace routing
+{
+namespace turns
+{
+namespace sound
+{
+
+long FindInStrArray(std::u32string_view haystack, char32_t needle)
+{
+  auto begin = haystack.begin();
+  for (auto i = begin; i != haystack.end(); ++i)
+  {
+    const strings::UniChar item = *i;
+    if (item == needle)
+      return (long)i;
+  }
+  return -1;
+}
+
+void HungarianBaseWordTransform(std::string & hungarianString)
+{
+  std::pair<std::string_view, std::string_view> constexpr kToReplace[] = {
+      {"e", "é"}, {"a", "á"}, {"ö", "ő"}, {"ü", "ű"}};
+  for (auto [base, harmonized] : kToReplace)
+  {
+    if (strings::EndsWith(hungarianString, base))
+    {
+      hungarianString.replace(hungarianString.size() - base.size(), base.size(), harmonized);
+      break;
+    }
+  }
+}
+
+bool EndsInAcronymOrNum(strings::UniString const & myUniStr)
+{
+  bool allUppercaseNum = true;
+  strings::UniString lowerStr = strings::MakeLowerCase(myUniStr);
+  for (int i = myUniStr.size() - 1; i > 0; i--)
+  {
+    // if we've reached a space, we're done here
+    if (myUniStr[i] == ' ')
+      break;
+    // we've found a char that is already lowercase and not a number,
+    // therefore the string is not exclusively uppercase/numeric
+    else if (myUniStr[i] == lowerStr[i] && !std::isdigit(myUniStr[i]))
+    {
+      allUppercaseNum = false;
+      break;
+    }
+  }
+  return allUppercaseNum;
+}
+
+uint8_t CategorizeHungarianAcronymsAndNumbers(std::string const & hungarianString)
+{
+  constexpr std::array<std::string_view, 14> backNames = {
+      "A",  // a
+      "Á",  // á
+      "H",  // há
+      "I",  // i
+      "Í",  // í
+      "K",  // ká
+      "O",  // o
+      "Ó",  // ó
+      "U",  // u
+      "Ű",  // ú
+      "0",  // nulla or zéró
+      "3",  // három
+      "6",  // hat
+      "8",  // nyolc
+  };
+
+  constexpr std::array<std::string_view, 31> frontNames = {
+      // all other letters besides H and K
+      "B", "C", "D", "E", "É", "F", "G", "J", "L", "M", "N", "Ö", "Ő",
+      "P", "Q", "R", "S", "T", "Ú", "Ü", "V", "W", "X", "Y", "Z",
+      "1",  // egy
+      "2",  // kettő
+      "4",  // négy
+      "5",  // öt
+      "7",  // hét
+      "9",  // kilenc
+  };
+
+  constexpr std::array<std::string_view, 5> specialCaseFront = {
+      "10",  // tíz special case front
+      "40",  // negyven front
+      "50",  // ötven front
+      "70",  // hetven front
+      "90",  // kilencven front
+  };
+
+  constexpr std::array<std::string_view, 4> specialCaseBack = {
+      "20",  // húsz back
+      "30",  // harminc back
+      "60",  // hatvan back
+      "80",  // nyolcvan back
+  };         // TODO: consider regex
+
+  //'100', // száz back
+
+  for (size_t i = hungarianString.length(); i > 0; i--)
+  {
+    std::string hungarianSub = hungarianString.substr(0, i);
```

Comment:
No it actually works fine in this case because we're looking for substring matches against other small strings and we return on the first match... if a string ends in multi-byte characters, we'll just loop until we match one of the strings above. I have test cases for this, it currently works as intended

Reply:
Comparing partial byte sequences instead of characters is possible, and even may work in most cases. This is not the right way to work with characters. If the code was iterating using characters, not bytes, then it was safe.

Reply:
@biodranik tests pass with simply calling `.endsWith()` repeatedly so if we're lucky that's all that's needed